File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        arr = []
        def dfs(node):
            if not node or len(arr) == k:
                return
            if not node.left and not node.right:
                arr.append(node.val)
                return
            dfs(node.left)
            arr.append(node.val)
            dfs(node.right)
            
        dfs(root)
        
        return arr[k - 1]
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    // maybe only when it happens for either good or bad
    // that we can imagine that possiblities beyond the current
    // even exists.
    diameterOfBinaryTree(root) {
        let maxDiameter = 0;
        function walk(node) {
            if (!node) return 0; // this is different from the return type, here null return to node hence answer is 0
            const leftDepth = walk(node.left);
            const rightDepth = walk(node.right);
            
            maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth);
            
            // here node return to node therefore 
            // that path counts as 1, which is different from null -> node
            return Math.max(leftDepth, rightDepth) + 1; 
        };

        walk(root);
        return maxDiameter;
    }
    
}
